src/lib/peers/peer-connection/PeerConnection.ts

```diff
@@ -0,0 +1,239 @@
+
+class PeerConnection {
+  private publicKey: string;
+  private keyManager: KeyManager;
+  private peerManager: PeerManager;
+
+  private peerClient: PeerClient;
+  private connected: boolean = false;
+  private credentials: grpc.ChannelCredentials;
+
+  constructor(publicKey: string, keyManager: KeyManager, peerManager: PeerManager) {
+    this.publicKey = publicKey;
+    this.keyManager = keyManager;
+    this.peerManager = peerManager;
+
+    const pkiInfo = keyManager.PKIInfo;
+    if (pkiInfo.caCert && pkiInfo.cert && pkiInfo.key) {
+      this.credentials = grpc.credentials.createSsl(pkiInfo.caCert, pkiInfo.key, pkiInfo.cert);
+    } else {
+      this.credentials = grpc.credentials.createInsecure();
+    }
+  }
+
+  private async connectDirectly(): Promise<PeerClient> {
+    // try to create a direct connection
+    if (this.getPeerInfo().connectedAddress) {
+      // direct connection attempt
+      const address = this.getPeerInfo().connectedAddress!;
+      const peerClient = new PeerClient(address.toString(), this.credentials);
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().connectedAddress) {
+      throw Error('peer does not have a connected address');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  private async connectHolePunch(): Promise<PeerClient> {
+    // try to hole punch to peer via relay peer
+    if (!this.connected && this.getPeerInfo().relayPublicKey) {
+      // connect to relay and ask it to create a relay
+      console.log('requesting udp hole punch connection');
+      const connectedAddress = await this.peerManager.turnClient.requestHolePunchConnection(
+        this.getPeerInfo().relayPublicKey!,
+        this.getPeerInfo().publicKey,
+      );
+      const peerClient = new PeerClient(connectedAddress.toString(), this.credentials);
+
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().relayPublicKey) {
+      throw Error('peer does not have relay public key specified');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  private async connectRelay(): Promise<PeerClient> {
+    // try to relay to peer via relay peer
+    if (!this.connected && this.getPeerInfo().relayPublicKey) {
+      // turn relay
+      // connect to relay and ask it to create a relay
+      const connectedAddress = await this.peerManager.turnClient.requestPeerConnection(
+        this.getPeerInfo().publicKey,
+        this.getPeerInfo().relayPublicKey!,
+      );
+      const peerClient = new PeerClient(connectedAddress.toString(), this.credentials);
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().relayPublicKey) {
+      throw Error('peer does not have relay public key specified');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  async connectFirstChannel() {
```

Comment:
Priority of connection should just be encoded in the code. No need to call it `FirstChannel`. Obviously we prefer direct, then hole punch then relay.

Reply:
Actually the way it works at the moment is all channels are tried in parallel, we can do this because the quickest to slowest order in terms of connection speed is the same as the order you mentioned; direct -> hole punch -> relay. It just saves us time not to have to wait to try each before going on to the second one.

Reply:
Also, I have put the connection logic into the `connect` method to eliminate the redundant `connectFirstChannel` method.